endterm/fastApiProject/main.py

```diff
@@ -0,0 +1,430 @@
+def get_db():
+    db = SessionLocal()
+    try:
+        yield db
+    finally:
+        db.close()
+
+def get_current_time():
+    return datetime.datetime.now()
+
+def generate_log_id():
+    return uuid.uuid4()
+
+def create_log(filename: str, username: str, method, message: str):
+    log_to_file.delay(f"{filename}", f"LOG_ID: {generate_log_id()} ---- "
+                                                 f"TIME: {get_current_time()} ---- "
+                                                 f"USER: {username} ---- "
+                                                 f"OPERATION: {method} ---- "
+                                                 f"MESSAGE: {message}")
+# User registration endpoint
+@app.post("/register", response_model=schemas.User, tags=["users"])
+async def register(request: Request, user: schemas.UserCreate, db: Session = Depends(get_db)):
+    message = f"SOME USER TRYING TO REGISTER. REGISTRATION DATA: {user.__str__()}"
+    create_log(user_log_filename, user.username, request.method, message)
+    db_user = user_crud.get_user_by_email(db, email=user.email)
+    if db_user:
+        raise HTTPException(status_code=400, detail="Email already registered")
+    message = f"{user.username} REGISTERED SUCCESSFULLY"
+    create_log(user_log_filename, user.username, request.method, message)
+    return user_crud.create_user(db=db, user=user)
+
+
+# Login endpoint
+@app.post("/token", response_model=Token, tags=["users"])
+async def login(request: Request, form_data: OAuth2PasswordRequestForm = Depends(), db: Session = Depends(get_db)):
+    user = user_crud.authenticate_user(db, form_data.username, form_data.password)
+    message = f"{user.username} IS TRYING TO LOG IN"
+    create_log(user_log_filename, user.username, request.method, message)
+    if not user:
+        raise HTTPException(status_code=400, detail="Incorrect username or password")
+    access_token_expires = timedelta(minutes=user_crud.ACCESS_TOKEN_EXPIRE_MINUTES)
+    refresh_token_expires = timedelta(minutes=user_crud.REFRESH_TOKEN_EXPIRE_MINUTES)
+    access_token = user_crud.create_access_token(
+        data={"sub": user.username}, expires_delta=access_token_expires
+    )
+    refresh_token = user_crud.create_refresh_token(
+        data={"sub": user.username}, expires_delta=refresh_token_expires
+    )
+    result = Token(access_token=access_token, refresh_token=refresh_token)
+    message = f"{user.username} SUCCESSFULLY LOGGED IN"
+    create_log(user_log_filename, user.username, request.method, message)
+    return result
+
+
+@app.get("/countries/", response_model=list[schemas.Country], tags=["country"])
+def read_countries(request: Request, skip: int = 0, limit: int = 10, db: Session = Depends(get_db), username: str = Depends(TokenVerifier.verify_token)):
+    message = f"{username} IS READING ALL COUNTRIES"
+    countries = country_crud.get_countries(db, skip=skip, limit=limit)
+    create_log(country_log_filename, username, request.method, message)
```

Comment:
https://www.geeksforgeeks.org/decorators-with-parameters-in-python/

Reply:
Done